ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/modalActivityBlock/modalActivityBlock.component.ts

```diff
@@ -0,0 +1,165 @@
+    @ViewChild('delete_button', {read: ElementRef}) private deleteButtonRef: ElementRef;
+
+    public activityForm: ActivityForm;
+
+    constructor(private readonly activityServiceAgent: ActivityServiceAgent,
+                private dialog: MatDialog,
+                private cdr: ChangeDetectorRef) {
+    }
+
+    public deleteActivityInstance(): void {
+        this.activityServiceAgent.deleteActivityInstance(
+            this.studyGuid,
+            this.instance.instanceGuid
+            )
+            .pipe(take(1))
+            .subscribe(() => {
+                this.deleteActivity.emit(this.instance.instanceGuid);
+            });
+    }
+
+    public openEditDialog(): void {
+        this.getFullActivity()
+            .then((activity: ActivityForm) => {
+                this.activityForm = activity;
+                this.openDialog(this.editModalRef, EDIT_DIALOG_CONFIG, this.closeEditDialog.bind(this));
+            })
+            .catch((err) => {
+                console.error('An error during getting a full activity', err);
+            });
+    }
+
+    public closeEditDialog(): void {
+        this.getActivityInstance()
+            .then(() => {
+                this.cdr.detectChanges();
+                this.dialog.closeAll();
+            })
+            .catch((err) => {
+                console.error('An error during getting an activity instance', err);
+            });
+    }
+
+    public openDeleteDialog(): void {
+        this.openDialog(this.deleteModalRef, this.getDeleteDialogConfig(), this.closeEditDialog.bind(this));
+    }
+
+    private getFullActivity(): Promise<ActivityForm> {
+        return new Promise((resolve, reject) => {
+            this.getActivity$()
+                .subscribe((activity: ActivityForm) => {
+                    resolve(activity);
+                }, (err) => {
+                    reject(err);
+                });
+        });
+    }
+
+    private getActivity$(): Observable<ActivityForm> {
+        return this.activityServiceAgent.getActivity(
+            of(this.studyGuid),
+            of(this.instance.instanceGuid)
+            )
+            .pipe(take(1));
+    }
+
+
+    private getActivityInstance(): Promise<void> {
+        return new Promise((resolve, reject) => {
+            this.activityServiceAgent.getActivitySummary(this.studyGuid, this.instance.instanceGuid)
+                .pipe(take(1))
+                .subscribe(activityInstance => {
+                    this.instance = activityInstance;
+                    resolve();
+                }, (err) => {
+                    reject(err);
+                });
+        });
+    }
+
+    private openDialog(templateRef: TemplateRef<any>, config: any, closeDialogCallback: (...args) => void): void {
+        const dialogRef = this.dialog.open(templateRef, config);
+
+        if (closeDialogCallback) {
+            dialogRef.beforeClosed().subscribe(result => {
+                closeDialogCallback(result);
```

Comment:
close is only going to happen once, so a pipe(take(1)) to dispose of subscription ought to be fine.